### PR TITLE
[autolinking] Introduce universal "apple" platform

### PR DESCRIPTION
**Pull request by tsapeta**
# Why

There is no reason to differentiate `ios`, `macos`, `tvos` platforms in autolinking and expo module configs. Having a universal `apple` platform is enough and simplifies a few things.

# How

- Now autolinking CLI supports only `apple` and `android` values for `--platform` option. It makes the TypeScript side closer to be Apple-platform-agnostic. `--platform apple` returns a list of modules that support any of iOS/macOS/tvOS platform.
- The responsibility for filtering out packages that are incompatible with a given target is moved to Ruby scripts where we already read the podspec from which we get the concrete platforms supported by the pod. In other words, podspec becomes the only source of truth about which Apple platforms are supported (no need to define this twice in the podspec and `expo-module.config.json`).
- Introduced new `generate-modules-provider` command in the CLI, with a slightly different semantic than the existing `generate-package-list`. Since autolinking can't read podspecs from TypeScript, we need to provide a list of packages that will be included in the generated modules provider, using the dedicated `--packages <packages...>` option.
- Updated all `expo-module.config.json` that already had `macos` or `tvos` platform. No need to update others yet – this is good for checking backwards compatibility.

# Test Plan

I ran `pod install` in `apps/bare-expo/ios`, `apps/bare-expo/macos`, `apps/expo-go/ios` – there are no changes to Podfile locks, which means that the same modules are being linked as before.

### CODE REVIEW

**Comment by fobos531** [comment]
How does this work with existing Expo Modules that were built up until now? Does it require any migration work?

File: packages/expo-module-template/expo-module.config.json [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
-  "platforms": ["ios", "android", "web"],
+  "platforms": ["apple", "android", "web"],
   "ios": {
```
Comment (reichhartd) [inline packages/expo-module-template/expo-module.config.json:3]:
```suggestion
  "apple": {
```

File: packages/expo-module-template-local/expo-module.config.json [hunk 2]
```diff
@@ -1,5 +1,5 @@
 {
-  "platforms": ["ios", "android", "web"],
+  "platforms": ["apple", "android", "web"],
   "ios": {
```
Comment (reichhartd) [inline packages/expo-module-template-local/expo-module.config.json:3]:
```suggestion
  "apple": {
```

**Comment by tsapeta** [comment]
@fobos531
No, it doesn't require any migration. This is a backwards compatible change, meaning that the existing `ios` platform is effectively the same as the new `apple`. In this PR I intentionally left other SDK packages unchanged, which confirms backwards compatibility.

File: packages/expo-module-template-local/expo-module.config.json (hunk 2 above)
Comment (tsapeta) [reply packages/expo-module-template-local/expo-module.config.json:3]: Good catch 🙈 Thanks!

**Review by Kudo — APPROVED** [review]
🔥

**Review by douglowder — APPROVED** [review]

File: packages/expo-modules-autolinking/src/ExpoModuleConfig.ts [hunk 3]
```diff
@@ -17,52 +22,67 @@ export class ExpoModuleConfig {
...
   supportsPlatform(platform: SupportedPlatform): boolean {
-    return this.rawConfig.platforms?.includes(platform) ?? false;
+    const supportedPlatforms = this.rawConfig.platforms ?? [];
+
+    if (platform === 'apple') {
+      // Apple platform is supported when any of iOS, macOS and tvOS is supported.
+      return supportedPlatforms.some((supportedPlatform) => {
+        return ['apple', 'ios', 'macos', 'tvos'].includes(supportedPlatform);
+      });
+    }
```
Comment (brentvatne) [inline packages/expo-modules-autolinking/src/ExpoModuleConfig.ts:32]: opened the diff looking for this specifically :+1: good to ensure that there is compat with existing configs

**Review by brentvatne — APPROVED** [review]

**Comment by libcthorne** [comment]
This change looks great, but I've encountered one issue.

When using Expo 49 (or any version <50), running `npx create-expo-module` seems to use the new `apple` identifier whereas older Expo is still looking for `ios`.

I think this is because `create-expo-module` has its own environment (even when run within an existing Expo project folder) and is pulling in the newer Expo code:
```
% npx create-expo-module       
Need to install the following packages:
create-expo-module@0.6.7
Ok to proceed? (y) 
```

I hit this issue while going through https://docs.expo.dev/modules/get-started/#creating-the-local-expo-module as iOS was failing to import the module.

I thought it worth mentioning in case there's a quick fix that could be added in the dependencies to avoid others encountering the same.

**Comment by tsapeta** [comment]
Hey @libcthorne,

You're right, I think I've changed `ios` to `apple` in the module template too early which caused it not being handled properly by the old version of autolinking (before SDK 50). I'll follow up on that and change it back to `ios` in the template.

I'm working on a guide how to add support for macOS and tvOS platforms to the expo module and it seems to be a good place to mention that if you want to support more than iOS, you should just change it to `apple`. Otherwise it's ok to use `ios` (it will never be dropped).

Sorry for the trouble!

**Pull request merged**